# Часть 1: Задания, решаемые без использования Python
# Задание 1
Исследуйте систему векторов на линейную зависимость или независимость без использования Python (вручную):

$\vec{a}_{1}=(1 ; 8 ;-1), \vec{a}_{2}=(-2 ; 3 ; 3), \vec{a}_{3}=(4 ;-11 ; 9)$

Подробно распишите решение, а в итоге сделайте вывод про то, являются ли векторы зависимыми или независимыми.

*Рассмотрим матричное представление векторов:*

$$
A = 
\begin{pmatrix}
1 & 8 & -1 \\
-2 & 3 & 3 \\
4 & -11 & 9
\end{pmatrix}
$$

*Векторы линейно зависимы, если определитель матрицы A равен нулю*

$$
\det(A) = 1*\det\begin{vmatrix}
3 & 3 \\
-11 & 9
\end{vmatrix} 
- 8*\det\begin{vmatrix}
-2 & 3 \\
4 & 9
\end{vmatrix} 
+ (-1)*\det\begin{vmatrix}
-2 & 3 \\
4 & -11
\end{vmatrix}
= 1*((3*9) - (3*(-11)))-8*((-2*9) - (3*4))+(-1)*((-2*-11) - (3*4)) = 290
$$ 

*__Ответ__: Так как определитель матрицы не равен нулю, то данные векторы являются линейно независимыми. Это означает, что ни один из векторов нельзя представить как линейную комбинацию других векторов системы.*

# Задание 2
При каких значениях параметра a система векторов является линейно зависимой? Найдите решение без использования Python (вручную).

$\vec{a}_{1}=(1 ; 0 ; 2 ;-3), \vec{a}_{2}=(2 ; 2 ;-4 ; a), \vec{a}_{3}=(3 ; 1 ; 2 ;-5)$


$$
A = 
\begin{pmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & a \\
-3 & a & -5
\end{pmatrix}
$$


*Векторы линейно зависимы, если определитель матрицы A равен нулю*

*Рассмотрим матричное представление векторов:*
$ A = \begin{pmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & 2 \\
-3 & a & -5
\end{pmatrix} $

*Чтобы найти условие линейной зависимости, рассмотрим определитель подматрицы из первых трех строк и столбцов:*  
$ \det(A_{3 \times 3}) = \begin{vmatrix}
1 & 2 & 3 \\
0 & 2 & 1 \\
2 & -4 & 2
\end{vmatrix} $

*Вычисляем определитель:*  
$ \det(A_{3 \times 3}) = 1*\begin{vmatrix} 2 & 1 \\ -4 & 2 \end{vmatrix} - 2*\begin{vmatrix} 0 & 1 \\ 2 & 2 \end{vmatrix} + 3*\begin{vmatrix} 0 & 2 \\ 2 & -4 \end{vmatrix} = 1*((2*2) - (-4*1)) - 2*((0*2) - (1*2)) + 3*((0*(-4)) - (2*2)) = 0$

*__Ответ:__ Поскольку определитель подматрицы $3 \times 3$ равен нулю, система векторов линейно зависима при любом значении параметра $a$.*

# Задание 3

Выразите определитель матрицы. Найдите решение без использования Python (вручную).


$
\left|\begin{array}{rrrr}1 & -1 & a & -1 \\ -1 & -2 & b & -1 \\ -2 & 0 & c & 1 \\ 0 & 1 & d & 0\end{array}\right|$

1) *Разложим определитель по элементам четвёртой строки:*
    $$
    \det(A) = 0*\begin{vmatrix} -1 & a & -1 \\ -2 & b & -1 \\ 0 & c & 1 \end{vmatrix}
    - 1*\begin{vmatrix} 1 & a & -1 \\ -1 & b & -1 \\ -2 & c & 1 \end{vmatrix}
    + d*\begin{vmatrix} 1 & -1 & -1 \\ -1 & -2 & -1 \\ -2 & 0 & 1 \end{vmatrix}
    - 0*\begin{vmatrix} 1 & -1 & a \\ -1 & -2 & b \\ -2 & 0 & c \end{vmatrix}
    =
    - \begin{vmatrix} 1 & a & -1 \\ -1 & b & -1 \\ -2 & c & 1 \end{vmatrix}
    + d*\begin{vmatrix} 1 & -1 & -1 \\ -1 & -2 & -1 \\ -2 & 0 & 1 \end{vmatrix}
    $$

2) *Найдем определитель для каждой матрицы меньшего порядка:*

    2.1) *Первая матрица*  
    $$det \begin{vmatrix} 1 & a & -1 \\ -1 & b & -1 \\ -2 & c & 1 \end{vmatrix}$$

    *Разложим по первой строке:*

    $$
    \det = 1*\begin{vmatrix} b & -1 \\ c & 1 \end{vmatrix} - a*\begin{vmatrix} -1 & -1 \\ -2 & 1 \end{vmatrix} + (-1)*\begin{vmatrix} -1 & b \\ -2 & c \end{vmatrix}
    =
    1 * (b*1 - (-1)*c) - a * ((-1)*1 - (-1)*(-2)) + (-1) * (-1)*(c - b*(-2))
    =
    -b + 2c + 3a
    $$

    2.2) *Вторая матрица*  
    $$det \begin{vmatrix} 1 & -1 & -1 \\ -1 & -2 & -1 \\ -2 & 0 & 1 \end{vmatrix}$$

    *Разложим по первой строке:*

    $$
    \det = 1 * \begin{vmatrix} -2 & -1 \\ 0 & 1 \end{vmatrix} - (-1) * \begin{vmatrix} -1 & -1 \\ -2 & 1 \end{vmatrix} + (-1) * \begin{vmatrix} -1 & -2 \\ -2 & 0 \end{vmatrix} = 1*((-2)*1 - (-1)*0) - (-1)*((-1)*1 - (-1)*(-2)) + (-1)*((-1)*0 - (-2)*(-2)) = -1
    $$

3) *Подставим в формулу:*  

    $-(-b + 2c + 3a) + d(-1) = b - 2c - 3a - d$


*__Ответ:__*
$
\det(A) = b - 2c - 3a - d
$


# Задание 4
Найдите значения параметров a, b и c, при которых матрицы A и B являются обратными. Найдите решение без использования Python (вручную).

$A=\left(\begin{array}{ccc}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2 c
\end{array}\right), B=\left(\begin{array}{rrr}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{array}\right)$

*Чтобы матрицы $ A $ и $ B $ были обратными, их произведение должно быть равно единичной матрице $I$.*

*Рассчитаем произведение $ AB $:*

$$
AB = \begin{pmatrix}
a & -2 & -1 \\
-15 & b+20 & 12 \\
10 & -19 & 2c
\end{pmatrix} 
\begin{pmatrix}
-4 & 3 & 5 \\
0 & 2 & 3 \\
-5 & -1 & -1
\end{pmatrix}
$$

1) *Первая строка:*

    1.1) Элемент 1: $a*(-4) + (-2)*0 + (-1)*(-5) = -4a + 5$  
    1.2) Элемент 2: $a*3 + (-2)*2 + (-1)*(-1) =  3a - 3$  
    1.3) Элемент 3: $a*5 + (-2)*3 + (-1)*(-1) = 5a - 5 $  

2) *Элементы второй строки:*

    2.1) Элемент 1: $-15*(-4) + (b+20)*0 + 12*(-5) = 0$  
    2.2) Элемент 2: $-15*3 + (b+20)*2 + 12*(-1) = 2b - 17$  
    2.3) Элемент 3: $-15*5 + (b+20)*3 + 12*(-1) = 3b - 27$  

1) *Элементы третьей строки:*

    3.1) Элемент 1: $10*(-4) + (-19)*0 + 2c*(-5) = -40 - 10c$  
    3.2) Элемент 2: $10*3 + (-19)*2 + 2c*(-1) = -8 - 2c$  
    3.3) Элемент 3: $10*5 + (-19)*3 + 2c*(-1) = -7 - 2c$  

*Приравняем каждый элемент произведения матриц AB к соответвующему элементу единичной матрицы и получим уравнения:*

1. $ -4a + 5 = 1 $
2. $ 3a - 3 = 0 $
3. $ 5a - 5 = 0 $
4. $ 2b - 17 = 1 $
5. $ 3b - 27 = 0 $
6. $ -40 - 10c = 0 $
7. $ -8 - 2c = 0 $
8. $ -7 - 2c = 1 $

*Решаем систему:*

1. Из уравнения $ -4a + 5 = 1 \Rightarrow -4a = -4 \Rightarrow a = 1 $
2. Из уравнения $ 2b - 17 = 1 \Rightarrow 2b = 18 \Rightarrow b = 9 $
3. Из уравнения $ -8 - 2c = 0 \Rightarrow -2c = 8 \Rightarrow c = -4 $

*Ответ: a = 1, b = 9, c = -4*  
*Эти значения удовлетворяют условию $ AB = I $, следовательно, матрицы $ A $ и $ B $ являются обратными*


# Задание 5
Найдите собственные векторы и собственные значения матрицы. Найдите решение без использования Python (вручную).


$\left(\begin{array}{rrr}4 & 0 & -1 \\ 2 & 2 & 1 \\ -1 & 0 & 4\end{array}\right)$

1) Найдем $A - \lambda I$.

$
A - \lambda I = \begin{pmatrix} 4 - \lambda & 0 & -1 \\ 2 & 2 - \lambda & 1 \\ -1 & 0 & 4 - \lambda \end{pmatrix}
$

2) Найдем определитель:

$
\det(A - \lambda I) = (4 - \lambda) \cdot \det\begin{pmatrix} 2 - \lambda & 1 \\ 0 & 4 - \lambda \end{pmatrix} - 0 + (-1) \cdot \det\begin{pmatrix} 2 & 1 \\ -1 & 4 - \lambda \end{pmatrix}= - \lambda^3 +10\lambda^2-31\lambda+30 = 
-(\lambda-2)(\lambda^2-8\lambda+15)=-(\lambda-2)(\lambda-3)(\lambda-5)=0
$

Получили три собственных значения: $\lambda=2,\lambda=3,\lambda=5$

3) Найдем собственные вектора соответсвующие собственным значениям:

    **3.1) Для $\lambda = 2$:**
    
    Для $\lambda = 2$, подставляем в $A - \lambda I$:

    $
    A - 2I = \begin{pmatrix} 4 & 0 & -1 \\ 2 & 2 & 1 \\ -1 & 0 & 4 \end{pmatrix} - 2 \cdot \begin{pmatrix} 1 & 0 & 0 \\ 0 & 1 & 0 \\ 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} 2 & 0 & -1 \\ 2 & 0 & 1 \\ -1 & 0 & 2 \end{pmatrix}
    $

    Теперь решаем уравнение $(A - 2I)v = 0$.

    Составляем систему уравнений:

    $
    \begin{cases}
    2x - z = 0 \\
    2x + z = 0 \\
    -x + 2z = 0
    \end{cases}
    $

    Решаем систему:  
    Из первого уравнения: $z = 2x$.  
    Подставляем $z = 2x$ во второе уравнение: $2x + 2x = 0 \Rightarrow 4x = 0 \Rightarrow x = 0$.  
    Подставляем $x = 0$ в $z = 2x: z = 0$.  
    $y$ остаётся свободным параметром.

    Следовательно, собственный вектор:

    $
    v_1 = \begin{pmatrix} 0 \\ 1 \\ 0 \end{pmatrix}
    $

    **3.2) Для $ \lambda = 3 $:**

    Для $ \lambda = 3 $, подставляем в $ A - \lambda I $:

    $
    A - 3I = \begin{pmatrix} 4 & 0 & -1 \\ 2 & 2 & 1 \\ -1 & 0 & 4 \end{pmatrix} - 3 \cdot \begin{pmatrix} 1 & 0 & 0 \\ 0 & 1 & 0 \\ 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} 1 & 0 & -1 \\ 2 & -1 & 1 \\ -1 & 0 & 1 \end{pmatrix}
    $

    Теперь решаем уравнение $ (A - 3I)v = 0 $.

    Составляем систему уравнений:

    $
    \begin{cases}
    x - z = 0 \\
    2x - y + z = 0 \\
    -x + z = 0
    \end{cases}
    $

    Решаем систему:  
    Из первого уравнения: $ x = z $.  
    Из третьего уравнения: $ -x + z = 0 \Rightarrow x = z $ (совпадает с первым).  
    Подставляем $ x = z $ во второе уравнение: $ 2z - y + z = 0 \Rightarrow 3z - y = 0 \Rightarrow y = 3z $.  

    Итак, $ x = z $, $ y = 3z $, $z$ — свободный параметр.

    Собственный вектор:

    $
    v_2 = \begin{pmatrix} 1 \\ 3 \\ 1 \end{pmatrix}
    $

    **3.3) Для $ \lambda = 5 $:**

    Для $ \lambda = 5 $, подставляем в $ A - \lambda I $:

    $
    A - 5I = \begin{pmatrix} 4 & 0 & -1 \\ 2 & 2 & 1 \\ -1 & 0 & 4 \end{pmatrix} - 5 \cdot \begin{pmatrix} 1 & 0 & 0 \\ 0 & 1 & 0 \\ 0 & 0 & 1 \end{pmatrix} = \begin{pmatrix} -1 & 0 & -1 \\ 2 & -3 & 1 \\ -1 & 0 & -1 \end{pmatrix}
    $

    Теперь решаем уравнение $ (A - 5I)v = 0 $.

    Составляем систему уравнений:

    $
    \begin{cases}
    -x - z = 0 \\
    2x - 3y + z = 0 \\
    -x - z = 0
    \end{cases}
    $

    Решаем систему:  
    Из первого уравнения: $ x = -z $.  
    Третье уравнение идентично первому: $ x = -z $.  
    Подставляем $ x = -z $ во второе уравнение: $ 2(-z) - 3y + z = 0 \Rightarrow -2z - 3y + z = 0 \Rightarrow -z - 3y = 0 \Rightarrow y = -\frac{1}{3}z $.  

    Итак, $ x = -z $, $ y = -\frac{1}{3}z $, $z$ — свободный параметр. 

    Собственный вектор:

    $
    v_3 = \begin{pmatrix} -1 \\ -\frac{1}{3} \\ 1 \end{pmatrix}
    $




# Задание 6
Найдите сумму сингулярных чисел для матрицы:

$A=\left(\begin{array}{ccc}-2 & 0 & 1 \\ 0 & -2 & 1 \\ 1 & 1 & -1\end{array}\right)$

Для нахождения суммы сингулярных чисел матрицы $A$, необходимо вычислить собственные значения матрицы $A^T A$, а затем взять их квадратные корни

1) Вычислим произведение $A^T A$:

$$
A^T A = \begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix}
\begin{pmatrix}
-2 & 0 & 1 \\
0 & -2 & 1 \\
1 & 1 & -1
\end{pmatrix}
=
\begin{pmatrix}
4 + 0 + 1 & 0 + 0 + 1 & -2 + 0 - 1 \\
0 + 0 + 1 & 0 + 4 + 1 & 0 - 2 - 1 \\
-2 + 0 - 1 & 0 - 2 - 1 & 1 + 1 + 1
\end{pmatrix}
=
\begin{pmatrix}
5 & 1 & -3 \\
1 & 5 & -3 \\
-3 & -3 & 3
\end{pmatrix}
$$

2) Найдем собственные значения матрицы $A^T A$

    Для нахождения собственных значений, решим характеристическое уравнение:

    $
    \det(A^T A - \lambda I) = 0
    $

    Вычислим детерминант:

    $$
    \det(A^T A - \lambda I) = \det \begin{pmatrix}
    5 - \lambda & 1 & -3 \\
    1 & 5 - \lambda & -3 \\
    -3 & -3 & 3 - \lambda
    \end{pmatrix}
    =
    (5 - \lambda) \det \begin{pmatrix} 5 - \lambda & -3 \\ -3 & 3 - \lambda \end{pmatrix} - 1 \det \begin{pmatrix} 1 & -3 \\ -3 & 3 - \lambda \end{pmatrix} + (-3) \det \begin{pmatrix} 1 & 5 - \lambda \\ -3 & -3 \end{pmatrix}
    =
    -\lambda^3 + 13\lambda^2-36\lambda=-\lambda(\lambda^2-13\lambda+36) = -\lambda(\lambda-4)(\lambda-9)=0
    $$
    Следовательно:
    $
    \lambda_1 = 9,\lambda_2 = 4,\lambda_3 = 0
    $

3) Найдем сингулярные числа:

    Сингулярные числа это квадратные корни из собственных значений матрицы $A^T A$:

    $
    \sigma_1 = \sqrt{\lambda_1} = \sqrt{9} = 3
    $

    $
    \sigma_2 = \sqrt{\lambda_2} = \sqrt{4} = 2
    $

    $
    \sigma_3 = \sqrt{\lambda_3} = \sqrt{0} = 0
    $

4) Найдем сумму сингулярных чисел

    $$
    \sigma_1 + \sigma_2 + \sigma_3 = 3 + 2 + 0 = 5
    $$

    Ответ: Сумма сингулярных чисел равна 5


# Часть 2: Задания, решаемые с помощью Python
# Задание 7

Загрузите данные. Изучите, есть ли в них пропущенные значения?Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [20]:
import pandas as pd
import numpy as np

In [21]:
# Загружаем данные из файла
data = pd.read_csv("data.xls")

# Выводим первые пять записей
data.head()

season  yr  mnth  holiday  weekday  workingday  weathersit       temp  \
0       1   0     1        0        6           0           2  14.110847   
1       1   0     1        0        0           0           2  14.902598   
2       1   0     1        0        1           1           1   8.050924   
3       1   0     1        0        2           1           1   8.200000   
4       1   0     1        0        3           1           1   9.305237   

      atemp      hum  windspeed(mph)  windspeed(ms)   cnt  
0  18.18125  80.5833       10.749882       4.805490   985  
1  17.68695  69.6087       16.652113       7.443949   801  
2   9.47025  43.7273       16.636703       7.437060  1349  
3  10.60610  59.0435       10.739832       4.800998  1562  
4  11.46350  43.6957       12.522300       5.597810  1600

In [22]:
# Считаем пропущенные элементы по столбцам
data.isnull().sum()

season            0
yr                0
mnth              0
holiday           0
weekday           0
workingday        0
weathersit        0
temp              0
atemp             0
hum               0
windspeed(mph)    0
windspeed(ms)     0
cnt               0
dtype: int64

В данных нет пропущенных значений

# Задание 8
Исследуем данные на наличие линейной зависимости между столбцами. Найдите ранг матрицы данных. Сравните его с количеством столбцов, сделайте вывод. Если линейная зависимость есть - напишите, между какими столбцами и удалите один из них.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [23]:
# Рассчитаем корреляцию между столбцами
corr_matrix = data.corr()
corr_matrix

season        yr      mnth   holiday   weekday  workingday  \
season          1.000000 -0.001844  0.831440 -0.010537 -0.003080    0.012485   
yr             -0.001844  1.000000 -0.001792  0.007954 -0.005461   -0.002013   
mnth            0.831440 -0.001792  1.000000  0.019191  0.009509   -0.005901   
holiday        -0.010537  0.007954  0.019191  1.000000 -0.101960   -0.253023   
weekday        -0.003080 -0.005461  0.009509 -0.101960  1.000000    0.035790   
workingday      0.012485 -0.002013 -0.005901 -0.253023  0.035790    1.000000   
weathersit      0.019211 -0.048727  0.043528 -0.034627  0.031087    0.061200   
temp            0.334315  0.047604  0.220205 -0.028556 -0.000170    0.052660   
atemp           0.342876  0.046106  0.227459 -0.032507 -0.007537    0.052182   
hum             0.205445 -0.110651  0.222204 -0.015937 -0.052232    0.024327   
windspeed(mph) -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
windspeed(ms)  -0.229046 -0.011817 -0.207502  0.006292  0.014282   -0.018796   
cnt             0.406100  0.566710  0.279977 -0.068348  0.067443    0.061156   

                weathersit      temp     atemp       hum  windspeed(mph)  \
season            0.019211  0.334315  0.342876  0.205445       -0.229046   
yr               -0.048727  0.047604  0.046106 -0.110651       -0.011817   
mnth              0.043528  0.220205  0.227459  0.222204       -0.207502   
holiday          -0.034627 -0.028556 -0.032507 -0.015937        0.006292   
weekday           0.031087 -0.000170 -0.007537 -0.052232        0.014282   
workingday        0.061200  0.052660  0.052182  0.024327       -0.018796   
weathersit        1.000000 -0.120602 -0.121583  0.591045        0.039511   
temp             -0.120602  1.000000  0.991702  0.126963       -0.157944   
atemp            -0.121583  0.991702  1.000000  0.139988       -0.183643   
hum               0.591045  0.126963  0.139988  1.000000       -0.248489   
windspeed(mph)    0.039511 -0.157944 -0.183643 -0.248489        1.000000   
windspeed(ms)     0.039511 -0.157944 -0.183643 -0.248489        1.000000   
cnt              -0.297391  0.627494  0.631066 -0.100659       -0.234545   

                windspeed(ms)       cnt  
season              -0.229046  0.406100  
yr                  -0.011817  0.566710  
mnth                -0.207502  0.279977  
holiday              0.006292 -0.068348  
weekday              0.014282  0.067443  
workingday          -0.018796  0.061156  
weathersit           0.039511 -0.297391  
temp                -0.157944  0.627494  
atemp               -0.183643  0.631066  
hum                 -0.248489 -0.100659  
windspeed(mph)       1.000000 -0.234545  
windspeed(ms)        1.000000 -0.234545  
cnt                 -0.234545  1.000000

In [24]:
# Определим ранг матрицы
matrix_rank = np.linalg.matrix_rank(data)
print("Ранг матрицы:", matrix_rank)

Ранг матрицы: 12


In [25]:
# Сравним ранг с количеством столбцов
print(f'Количество столбцов в таблице: {data.shape[1]} \nРанг матрицы: {matrix_rank}')

Количество столбцов в таблице: 13 
Ранг матрицы: 12


Количество столбцов в таблице больше ранга матрицы, следовательно присутсвует линейная зависимость

In [28]:
corr_matrix[(corr_matrix > 0.7) & (corr_matrix < 1.0)]

season  yr     mnth  holiday  weekday  workingday  \
season              NaN NaN  0.83144      NaN      NaN         NaN   
yr                  NaN NaN      NaN      NaN      NaN         NaN   
mnth            0.83144 NaN      NaN      NaN      NaN         NaN   
holiday             NaN NaN      NaN      NaN      NaN         NaN   
weekday             NaN NaN      NaN      NaN      NaN         NaN   
workingday          NaN NaN      NaN      NaN      NaN         NaN   
weathersit          NaN NaN      NaN      NaN      NaN         NaN   
temp                NaN NaN      NaN      NaN      NaN         NaN   
atemp               NaN NaN      NaN      NaN      NaN         NaN   
hum                 NaN NaN      NaN      NaN      NaN         NaN   
windspeed(mph)      NaN NaN      NaN      NaN      NaN         NaN   
windspeed(ms)       NaN NaN      NaN      NaN      NaN         NaN   
cnt                 NaN NaN      NaN      NaN      NaN         NaN   

                weathersit      temp     atemp  hum  windspeed(mph)  \
season                 NaN       NaN       NaN  NaN             NaN   
yr                     NaN       NaN       NaN  NaN             NaN   
mnth                   NaN       NaN       NaN  NaN             NaN   
holiday                NaN       NaN       NaN  NaN             NaN   
weekday                NaN       NaN       NaN  NaN             NaN   
workingday             NaN       NaN       NaN  NaN             NaN   
weathersit             NaN       NaN       NaN  NaN             NaN   
temp                   NaN       NaN  0.991702  NaN             NaN   
atemp                  NaN  0.991702       NaN  NaN             NaN   
hum                    NaN       NaN       NaN  NaN             NaN   
windspeed(mph)         NaN       NaN       NaN  NaN             NaN   
windspeed(ms)          NaN       NaN       NaN  NaN             1.0   
cnt                    NaN       NaN       NaN  NaN             NaN   

                windspeed(ms)  cnt  
season                    NaN  NaN  
yr                        NaN  NaN  
mnth                      NaN  NaN  
holiday                   NaN  NaN  
weekday                   NaN  NaN  
workingday                NaN  NaN  
weathersit                NaN  NaN  
temp                      NaN  NaN  
atemp                     NaN  NaN  
hum                       NaN  NaN  
windspeed(mph)            1.0  NaN  
windspeed(ms)             NaN  NaN  
cnt                       NaN  NaN

Присутствует линейная зависимость между столбцами:

1) mnth и season: 0.83144
2) atemp и temp: 0.991702
3) windspeed(ms) и windspeed(mph): 1.0

Удалим столбец windspeed(mph), так как он соответсвует windspeed(ms) и они оба показывают скорость ветра но в разных единицах измерения

In [27]:
data = data.drop(columns=['windspeed(mph)'])

# Задание 9
Выведите корреляционную матрицу оставшихся переменных.

* Какая переменная сильнее всего коррелирует с целевой?
* Есть ли мультиколлинеарность? Между какими переменными? Если есть корреляция выше 0.9 - оставьте только 1 признак. Объясните выбор  
Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [32]:
data.corr()['cnt'].sort_values()

weathersit      -0.297391
windspeed(ms)   -0.234545
hum             -0.100659
holiday         -0.068348
workingday       0.061156
weekday          0.067443
mnth             0.279977
season           0.406100
yr               0.566710
temp             0.627494
atemp            0.631066
cnt              1.000000
Name: cnt, dtype: float64

Больше всего с целевой переменной cnt коррелируют temp(0.627494) и atemp(0.631066). Между этими двумя переменными наблюдается высокая корреляция (0.991702), что указывает на проблему мультиколлинеарности. Чтобы устранить её, оставим только atemp, так как он сильнее коррелирует с целевой переменной.

In [33]:
data = data.drop(columns=['temp'])

# Задание 10
Разделите данные на предикторы и целевую переменную. Обучите модель с помощью классической линейной регрессии. Линейную регрессию необходимо реализовать без готовых библиотек.

Оцените качество модели с помощью хотя бы двух метрик качества.

Распишите решение с помощью кода на языке Python и напишите текстом интерпретацию полученного вывода.

In [41]:
# Предикторы
X = data.drop(columns=["cnt"])

# Целевая переменная
y = data["cnt"]

# Добавляем единичный столбец
X = np.c_[np.ones(X.shape[0]), X]

# Преобразуем в numpy array для дальнейшей работы
X = np.array(X)
y = np.array(y)

# Находим коэффициенты линейной регрессии
beta = np.linalg.inv(X.T @ X) @ X.T @ y

# Формируем предсказание модели
y_pred = X @ beta

# Оцениваем метрики качества
mse = np.mean((y - y_pred) ** 2)
mape = np.mean(np.abs((y - y_pred) / y)) * 100

print(f'mse: {mse} \nmape: {mape}')

mse: 750791.39 
mape: 45.59411204818117


Модель имеет высокие ошибки, что говорит о том, что ее предсказания далеко от реальных значений